### PR TITLE
Revised FrameData and snapshot API, added integration tests.

### LINKED ISSUE #139: Make FrameData private and merge _RawFrameData and FrameData
## Description

From @bdice suggestions in [this](https://github.com/glotzerlab/garnett/pull/134#discussion_r367767209) comment in PR #138. 

> I am also unsure whether the FrameData class should remain a public class, or whether the Frame.frame_data attribute should remain public. I am uncertain what the rationale was for this design. From my experience with the garnett API, it would be sensible to make FrameData private as _FrameData and to make Frame.frame_data into Frame._frame_data. In that case, I would also remove the to_hoomd_snapshot and other user-facing methods in the _FrameData class, which would be purely an internal storage container. Notably, len(FrameData) returns the number of types while len(Frame) returns the number of particles. Longer-term, it may be a good strategy to unify the _RawFrameData and _FrameData classes -- I can't see the reason for this class structure.

## Proposed solution
- Make `FrameData` private (refactor `FrameData` to `_FrameData`)
- Convert `_FrameData` to an internal storage container
    - Remove functions like `make_snapshot` and `to_hoomd_snapshot`/`copyto_snapshot`
- Merge `_FrameData` and `_RawFrameData`

### PR DESCRIPTION
**Pull request by bdice**
## Description
This completes some work begun in #144 and described further in #139. There are a few pieces to this PR, but I didn't feel confident about making the changes in #139 without also updating our CI to test those changes.

1. The `FrameData` class has been renamed to `_FrameData` and the `Frame.frame_data` attribute is now renamed to `Frame._frame_data`. **This breaks the API for FrameData and Frame.frame_data.**

2. I removed the private module-level methods for converting between HOOMD snapshots and `garnett.trajectory.Frame` objects. I replaced them with a `classmethod` constructor `Frame.from_hoomd_snapshot` and moved the `_to_hoomd_snapshot` implementation into the `Frame.to_hoomd_snapshot` method. **This does not break any additional APIs except for the methods already removed in _FrameData.** (Those APIs were previously broken by #144 and this just cleans up the implementation.)

3. I added integration tests to CI so that we know the snapshot features work as intended. Previously we had to run tests locally. See below for more info.

4. I also added integration with Codecov to measure our coverage, since it was only two more lines of code: https://codecov.io/gh/glotzerlab/garnett/tree/feature%2Fframedata-snapshot-api/garnett

## Motivation and Context
Resolves: #139. I don't think that merging `_FrameData` and `_RawFrameData` makes sense right now, but this PR makes all that logic internal and it could be refactored at any time.

## How Has This Been Tested?
I have updated the testing infrastructure to use `conda` and install HOOMD from `conda-forge` and `plato-draw` from PyPI (along with the `fresnel`, `matplotlib`, and `povray` backends). This means the "integration" tests on CI should now cover the entire package, and local testing is no longer necessary.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist:
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/garnett/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [x] I have updated the [credits](https://github.com/glotzerlab/garnett/blob/master/doc/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/garnett/blob/master/changelog.rst).

### CODE REVIEW

File: garnett/trajectory.py [hunk 1]
```diff
@@ -281,10 +261,14 @@ def _validate_input_array(self, value, dim, nelem=None, dtype=None):
             raise ValueError("Input array must be of shape (N, {}) where N is the number of particles.".format(nelem))
         return value
 
-    def _raw_frame_to_frame(self, raw_frame, dtype=None):
-        """Generate a frame object from a raw frame object."""
+    @classmethod
+    def _raw_frame_data_to_frame_data(self, raw_frame, dtype=None):
```
Comment (lyrivera) [inline garnett/trajectory.py]: `cls` instead of `self`

Comment (bdice) [reply garnett/trajectory.py]:
```suggestion
    def _raw_frame_data_to_frame_data(cls, raw_frame, dtype=None):
```

File: garnett/trajectory.py [hunk 2]
```diff
@@ -375,23 +359,79 @@ def __ne__(self, other):
                             details="Use to_hoomd_snapshot with no argument.")
     def make_snapshot(self):
         "Create a HOOMD-blue snapshot object from this frame."
-        self.load()
```
Comment (lyrivera) [inline garnett/trajectory.py]: Not clear to me why remove the `load` statement.

File: garnett/trajectory.py [hunk 3]
```diff
@@ -375,23 +359,79 @@ def __ne__(self, other):
...
+                pass
+        return snapshot
 
     @deprecation.deprecated(deprecated_in="0.7.0",
                             removed_in="0.8.0",
                             current_version=__version__,
                             details="Use to_hoomd_snapshot.")
     def copyto_snapshot(self, snapshot):
         "Copy this frame to a HOOMD-blue snapshot."
-        self.load()
```
Comment (lyrivera) [inline garnett/trajectory.py]: Same as above, why remove the `load` statement?

File: .circleci/config.yml [hunk 4]
```diff
@@ -34,34 +34,74 @@ jobs:
...
+    working_directory: ~/repo
+
+    steps:
+      - checkout
+
+      - run:
+          name: Install compilers
+          command: |
+            source activate base
+            conda install -c conda-forge gcc_linux-64 gxx_linux-64 --yes --quiet
```
Comment (lyrivera) [inline .circleci/config.yml]: What do we need these compilers for? fresnel and hoomd?

**Review by lyrivera — COMMENTED** [review]
Left a few comments/questions. Will approve once those have been addressed or cleared. Other than that, lgtm.

File: .circleci/config.yml (hunk 4 above)
Comment (bdice) [reply .circleci/config.yml]: They're for compiling the DCD reader, which is Cythonized into C++. By default this image has no compilers but the other image we use has compilers. 😋

File: garnett/trajectory.py (hunk 2 above)
Comment (bdice) [reply garnett/trajectory.py]: This is called by `self.to_hoomd_snapshot` so it's redundant to have it twice.

File: .circleci/config.yml (hunk 4 above)
Comment (lyrivera) [reply .circleci/config.yml]: Ah, that makes sense.

File: garnett/trajectory.py (hunk 3 above)
Comment (bdice) [reply garnett/trajectory.py]: This is called by `self.to_hoomd_snapshot` so it's redundant to have it twice.

File: garnett/trajectory.py [hunk 5]
```diff
@@ -323,13 +307,13 @@ def _raw_frame_to_frame(self, raw_frame, dtype=None):
...
-        return self.frame_data is not None
+        return self._frame_data is not None
 
     def load(self):
         "Load the frame into memory."
-        if self.frame_data is None:
+        if self._frame_data is None:
             logger.debug("Loading frame.")
-            self.frame_data = self._raw_frame_to_frame(self.read(), dtype=self._dtype)
+            self._frame_data = self._raw_frame_data_to_frame_data(self.read(), dtype=self._dtype)
```
[the remark this answers is not in the document]
Comment (bdice) [reply garnett/trajectory.py]: I revised this. The behavior doesn't even use `cls`, so it's not necessary for this to be a `classmethod`. Instead, I used a `staticmethod` and adapted the `CifFileReader` class (which defines its own version of the method).

References:
- https://stackoverflow.com/questions/692040/is-it-bad-form-to-call-a-classmethod-as-a-method-from-an-instance
- https://stackoverflow.com/questions/26788214/super-and-staticmethod-interaction/26807879

File: garnett/trajectory.py (hunk 3 above)
Comment (lyrivera) [reply garnett/trajectory.py]: Ah, I missed that. Good then

File: garnett/trajectory.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lyrivera) [reply garnett/trajectory.py]: Static method makes more sense. Thanks for the clarification.

**Review by lyrivera — APPROVED** [review]
LGTM, thank!

**Pull request merged**